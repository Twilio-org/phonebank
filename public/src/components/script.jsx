import React, { Component } from 'react';

export default class ScriptPage extends Component {
  constructor(props) {
    super(props);
    this.renderQuestions = this.renderQuestions.bind(this);
  }

  componentDidMount() {
    const { script_id } = this.props.script_id;
    this.props.fetchScript(script_id);
    this.props.fetchScriptQuestions(script_id);
  }

  renderQuestions() {
    const { questions } = this.props;
    return questions.map(question =>
      (<li key={question.id}>{question.title}
        <ul>
          <li>Description: {question.description}</li>
          <li>Type: {question.type}</li>
          <li>Responses: {question.responses ? question.responses : 'Text Paragraph'}</li>
        </ul>
      </li>));
  }

  render() {
    const { script_info, questions } = this.props;
    const { name, body, description } = script_info;
    return (
      <div>
        <h1>{script_info ? name : null} Script Details</h1>
        <h4>Body:</h4>
        {script_info ? (<p>{body}</p>) : null}
        <h4>Description:</h4>
        {script_info ? (<p>{description}</p>) : null}
        <h4>Questions:</h4>
        <ol>
          {questions ? this.renderQuestions() : null}
        </ol>
      </div>
    );
  }
}
