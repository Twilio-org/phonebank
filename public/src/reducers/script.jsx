export const defaultScriptQuestions = [];

export const SET_SCRIPT_QUESTIONS = 'SET_SCRIPT_QUESTIONS';

export function scriptQuestionsReducer(state = defaultScriptQuestions, action) {
  const { type, payload } = action;
  switch (type) {
    case SET_SCRIPT_QUESTIONS:
      return {
        ...state,
        questions: payload
      };
    default:
      return state;
  }
}
