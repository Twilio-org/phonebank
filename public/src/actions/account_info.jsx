import axios from 'axios';
import { SET_USER_ACCOUNT_INFO } from '../reducers/account_info';

export function setAccountInfo(user) {
  return {
    type: SET_USER_ACCOUNT_INFO,
    payload: user
  };
}

export function fetchUser(id) {
  return dispatch => axios.get(`/users/${id}`, {
    headers: { 'Authorization': ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((res) => {
    const userData = res.data;
    console.log(userData, 'on fetch user')
    return dispatch(setAccountInfo(userData));
  })
  .catch((err) => {
    const customError = {
      message: `error fetching user info in account_info action fetchUserInfo: ${err}`,
      name: 'user info post request from account_info component'
    };
    throw customError;
  });
}
