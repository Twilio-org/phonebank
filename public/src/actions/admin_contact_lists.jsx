import axios from 'axios';
import { SET_CAMPAIGN_FORM_CONTACT_LIST, SET_CURRENT_CONTACT_LIST } from '../reducers/admin_contact_lists';

export function setContactListOptions(contactLists) {
  return {
    type: SET_CAMPAIGN_FORM_CONTACT_LIST,
    payload: contactLists
  };
}

export function setCurrentContactList(contactListObj) {
  return {
    type: SET_CURRENT_CONTACT_LIST,
    payload: contactListObj
  };
}


export function fetchAllContactLists() {
  return dispatch => axios.get('/contactLists')
    .then((res) => {
      const contactLists = res.data;
      dispatch(setContactListOptions(contactLists));
      return contactLists;
    })
    .catch((err) => {
      const customError = {
        message: `error fetching contact lists: ${err}`,
        name: 'contact lists get request from campaignPage component'
      };
      throw customError;
    });
}
