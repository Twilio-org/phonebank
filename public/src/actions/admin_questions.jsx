import axios from 'axios';
import { destroy } from 'redux-form';
import { SET_QUESTIONS, SET_QUESTION_CURRENT } from '../reducers/admin_questions';

export function setQuestionList(questionList) {
  return {
    type: SET_QUESTIONS,
    payload: questionList
  };
}

export function setCurrentQuestion(questionDataObj) {
  return {
    type: SET_QUESTION_CURRENT,
    payload: questionDataObj
  };
}

export function fetchAllQuestions() {
  return dispatch => axios.get('/questions', {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((questions) => {
    const { data: questionsList } = questions;
    return dispatch(setQuestionList(questionsList.sort((a, b) => (a.id - b.id))));
  })
  .catch((err) => {
    const customError = {
      message: `error fetching questions: ${err}`,
      name: 'question get request from script_form component'
    };
    throw customError;
  });
}

export function fetchQuestion(id) {
  return dispatch => axios.get(`/questions/${id}`, {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((res) => {
    const { data: questionData } = res;
    // TO-DO Find a better way to handle this data
    questionData.responses = questionData.responses ? questionData.responses.split(',') : '';
    return dispatch(setCurrentQuestion(questionData));
  })
  .catch((err) => {
    const customError = {
      message: `error fetching question info in questions action fetchQuestion: ${err}`,
      name: 'question info get request'
    };
    throw customError;
  });
}

export function createQuestion(questionInfo, history) {
  let { responses } = questionInfo;
  const { title, description, type } = questionInfo;
  // TO-DO: Find a better way to handle question response data
  responses = responses ? Object.keys(responses).map(key => responses[key]).join(',') : '';

  return dispatch => axios.post('/questions', {
    title,
    description,
    type,
    responses
  })
  .then((res) => {
    history.goBack();
    dispatch(destroy('CreateQuestion'));
    return res;
  })
  .catch((err) => {
    const customError = {
      message: `error in creating new questions to db: ${err}`,
      name: 'createQuestion function error in actions/questions.jsx'
    };
    throw customError;
  });
}
