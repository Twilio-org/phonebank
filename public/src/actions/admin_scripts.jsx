import axios from 'axios';
import { SET_SCRIPTS, SET_SCRIPT_CURRENT } from '../reducers/admin_scripts';

export function setScriptsList(scriptList) {
  return {
    type: SET_SCRIPTS,
    payload: scriptList
  };
}

export function setCurrentScript(scriptDataObj) {
  return {
    type: SET_SCRIPT_CURRENT,
    payload: scriptDataObj
  };
}

export function fetchAllScripts() {
  return dispatch => axios.get('/scripts', {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((scripts) => {
    console.log(scripts, '##### succesfully grabed scrpts')
    const { data: scriptList } = scripts;
    return dispatch(setScriptsList(scriptList));
  })
  .catch((err) => {
    console.log('error fetching all scripts: ', err);
  });
}
