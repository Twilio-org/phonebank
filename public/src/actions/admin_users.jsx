import axios from 'axios';
import { SET_USERS, SET_USER_CURRENT } from '../reducers/admin_users';

export function setUserList(usersList) {
  return {
    type: SET_USERS,
    payload: usersList
  };
}

export function setCurrentUser(userObject) {
  return {
    type: SET_USER_CURRENT,
    payload: userObject
  };
}

export function fetchAllUsers(currentUserId) {
  return dispatch => axios.get('/users', {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((users) => {
    const { data: usersList } = users;
    const modifiedUsersList = usersList.map((userObject) => {
      const { is_admin, is_banned, is_active } = userObject;
      return {
        ...userObject,
        is_admin: JSON.stringify(is_admin),
        is_banned: JSON.stringify(is_banned),
        is_active: JSON.stringify(is_active)
      };
    });
    // remove the current user from the user's list... so they don't accidentally
    // change their status from admin to volunteer and have all the bugs happen
    const filteredUsersList = modifiedUsersList.filter((object) => {
      const { id } = object;
      return id !== currentUserId;
    });
    return dispatch(setUserList(filteredUsersList));
  })
  .catch(err => console.log('problem fetching all users from db in action "fetchAllUsers"', err));
}

export function adminUpdateUserInfo(id, target, newValue, currentUserId, fetchAll = fetchAllUsers) {
  const params = {};
  params[target] = JSON.stringify(newValue);
  return dispatch => axios.put(`/users/${id}/manage`, params,
    {
      headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
    }
  )
  .then(() => {
    dispatch(fetchAll(currentUserId));
  })
  .catch(err => console.log('error with user management update action: ', err));
}
