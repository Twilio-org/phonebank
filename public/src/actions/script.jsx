import axios from 'axios';
import { SET_SCRIPT_QUESTIONS } from '../reducers/script';
import { setCurrentScript } from './admin_scripts';

export function setScriptQuestions(questions) {
  return {
    type: SET_SCRIPT_QUESTIONS,
    payload: questions
  };
}

export function fetchScript(id) {
  return dispatch => axios.get(`/scripts/${id}`)
    .then((res) => {
      const { data: scriptData } = res;
      return dispatch(setCurrentScript(scriptData));
    })
    .catch((err) => {
      const customError = {
        message: `error fetching script info in script action fetchScript: ${err}`,
        name: 'script info get request from script component'
      };
      throw customError;
    });
}

export function fetchScriptQuestions(id) {
  return dispatch => axios.get(`/scripts/${id}/scriptQuestions/`)
    .then((res) => {
      const { data: scriptQuestions } = res;
      return dispatch(setScriptQuestions(scriptQuestions));
    })
    .catch((err) => {
      const customError = {
        message: `error fetching script questions in script action fetchScriptQuestions: ${err}`,
        name: 'script questions get request from script component'
      };
      throw customError;
    });
}
