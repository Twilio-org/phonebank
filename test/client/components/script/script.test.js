import React from 'react';
import { shallow } from 'enzyme';
import { fetchScript, fetchScriptQuestions } from '../../../../public/src/actions/script';
import fixtures from '../../client_fixtures';
import { checkObjectProps } from '../../client_test_helpers.js';
import ScriptPage from '../../../../public/src/components/script/script';

const { current_script,
        scriptQuestionsFixture: questions } = fixtures.scriptViewFixtures;

const props = {
  id: 1,
  current_script,
  questions,
  fetchScript,
  fetchScriptQuestions
};

describe('<ScriptPage />', () => {
  const { id, questions, current_script } = props;
  describe('Rendering of <ScriptPage />', () => {
    const wrapper = shallow(<ScriptPage {...props} />);
    it('should render one div element', () => {
      const divElements = wrapper.find('div').length;
      expect(divElements).toBe(1);
    });
    it('should render three h4 elements', () => {
      const h4Elements = wrapper.find('h4').length;
      expect(h4Elements).toBe(3);
    });
    it('should render one h1 element', () => {
      const h1Elements = wrapper.find('h1').length;
      expect(h1Elements).toBe(1);
    });
    it('should render one ol element', () => {
      const olElements = wrapper.find('ol').length;
      expect(olElements).toBe(1);
    });
  });
  describe('Expected props', () => {
    const wrapper = shallow(<ScriptPage {...props} />).instance();
    const expectedProps = Object.keys(props);
    it('should have the expected props', () => {
      const actualProps = wrapper.props;
      expect(checkObjectProps(expectedProps, actualProps)).toBe(true);
    });
  });
  describe('ScriptPage props and methods', () => {
    describe('props', () => {
      describe('script id prop', () => {
        it('should be defined', () => {
          expect(id).toBeDefined();
        });
        it('should have a value of 1', () => {
          expect(id).toBe(1);
        });
      });
      describe('current_script prop', () => {
        it('should be defined', () => {
          expect(current_script).toBeDefined();
        });
        it('should be an object', () => {
          expect(typeof current_script).toBe('object');
        });
        it('should have 3 properties - name, body, description', () => {
          const scriptInfoKeys = Object.keys(current_script);
          const scriptPropsAreCorrect = checkObjectProps(scriptInfoKeys, current_script);
          expect(scriptInfoKeys.length).toBe(3);
          expect(scriptPropsAreCorrect).toBe(true);
        });
      });
      describe('questions prop', () => {
        const firstQuestion = questions[0];
        it('should be defined', () => {
          expect(questions).toBeDefined();
        });
        it('should be an array of 1-5 question objects', () => {
          expect(questions.length).toBeLessThanOrEqual(5);
          expect(questions.length).toBeGreaterThanOrEqual(1);
          expect(Array.isArray(questions)).toBe(true);
          expect(typeof firstQuestion).toBe('object');
        });
        it('should contain question objects that have 4 properties - title, description, type, and responses', () => {
          const firstQuestionKeys = Object.keys(firstQuestion);
          const firstQuestionPropsAreCorrect = checkObjectProps(firstQuestionKeys, firstQuestion);
          expect(firstQuestionKeys.length).toBe(10);
          expect(firstQuestionPropsAreCorrect).toBe(true);
        });
      });
      describe('fetchScript prop', () => {
        it('should be defined', () => {
          expect(fetchScript).toBeDefined();
        });
        it('should be a function', () => {
          expect(typeof fetchScript).toBe('function');
        });
      });
      describe('fetchScriptQuestions prop', () => {
        it('should be defined', () => {
          expect(fetchScriptQuestions).toBeDefined();
        });
        it('should be a function', () => {
          expect(typeof fetchScriptQuestions).toBe('function');
        });
      });
    });
  });
});
