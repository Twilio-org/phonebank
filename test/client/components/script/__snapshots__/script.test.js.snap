// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScriptPage /> renders correctly 1`] = `
<div>
  <h1>
    ScriptName
     Script Details
  </h1>
  <h4>
    Body:
  </h4>
  <p>
    ScriptBody
  </p>
  <h4>
    Description:
  </h4>
  <p>
    ScriptDescription
  </p>
  <h4>
    Questions:
  </h4>
  <ol>
    <li>
      Age
      <ul>
        <li>
          Description: 
          What is your age?
        </li>
        <li>
          Type: 
          singleselect
        </li>
        <li>
          Responses: 
          0-10,11-20,21-30,31-40,41-50,51-60,61+
        </li>
      </ul>
    </li>
    <li>
      Hobbies
      <ul>
        <li>
          Description: 
          What are your hobbies?
        </li>
        <li>
          Type: 
          multiselect
        </li>
        <li>
          Responses: 
          swimming,running,biking,sleeping,eating,weaving
        </li>
      </ul>
    </li>
    <li>
      Ice Cream
      <ul>
        <li>
          Description: 
          How do you feel about ice cream?
        </li>
        <li>
          Type: 
          paragraph
        </li>
        <li>
          Responses: 
          Text Paragraph
        </li>
      </ul>
    </li>
  </ol>
</div>
`;
